src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -2426,27 +2426,8 @@ internal void ClearCache()
                 return;
             }
 
-            // Loop until the cache is successfully zero'd out.
-            do
-            {
-                // If the GCHandle doesn't wrap a cache yet, there's nothing more to do.
-                RuntimeTypeCache? existingCache = (RuntimeTypeCache?)GCHandle.InternalGet(m_cache);
-                if (existingCache is null)
-                {
-                    return;
-                }
-
-                // Create a new, empty cache to replace the old one and try to substitute it in.
-                var newCache = new RuntimeTypeCache(this);
-                if (ReferenceEquals(GCHandle.InternalCompareExchange(m_cache, newCache, existingCache), existingCache))
-                {
-                    // We were successful, so there's nothing more to do.
-                    return;
-                }
-
-                // We raced with someone else to initialize the cache.  Try again.
-            }
-            while (true);
+            // Set the GCHandle to null. The cache will be re-created next time it is needed.
```

Comment:
GCHandle is a weak GC handle, so the code is generally prepared for it to have null Target.

Reply:
Hmm. I'm glad this can be simplified (it wasn't clear to me whether this was safe), but that raises a question for me...

You'd originally objected to such clearing because it would turn reflection into a mutable model.  But if clearing this cache is sufficient for hot reload needs, then it was always possible for reflection to be mutable, if a GC occurred at the right time.  Which suggests to me there's actually something else somewhere that needs to be cleared as well for this to be successful?  And that this test is actually racy:
https://github.com/dotnet/runtime/blob/3cbbadee12cc95bd62c70786d5408a2277a21e0a/src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs#L19

Reply:
Ah, the test is ok because RuntimeMethodInfo stores a reference back to the cache.

I'm still not clear on my other question, though...

Reply:
Yes, there is a chance for EnC added method to be observable by reflection even today. Chances of it being observed in practice are low. It requires more than just GC happening at the right time. It would need to be something like:

1. Reflection method that returns MethodInfo is called.
2. All MethodInfos referencing affected type are released by the user code. Otherwise, the GC won't clear the weak GC handle since MethodInfos would be keeping the cache alive via `m_reflectedTypeCache` field.
3. Method is added by EnC
4. GC happens and clears the weak GC handle.
5. Reflection method that returns MethodInfo from step 1 is called. The result will include the EnC added method.

> And that this test is actually racy:

This test is ok, but the implementation overall is racy. For example, imagine there is a background task that does `foreach (Type t in ...) t.GetMethods()` in an infinite loop in your program. The following may happen:

1. Cache is cleared
2. Cache is re-created and populated by the background task
3. EnC method is added by the VM
4. The edit is not observable by reflection

I am not saying that this is a problem. It is in line with the overall level of EnC dependability.

Reply:
> The following may happen

Yup, absolutely. I was just under the impression that such races weren't possible with EnC today,  but they are,  just very rare. That makes me feel even less bad about the reliability of the cache clearing.